resolwe/permissions/models.py

```diff
@@ -0,0 +1,216 @@
+        # order of properties is important: properties listed sooner have
+        # precedence over properties listed later when inheriting permission
+        # group.
+        self._container_properties = ("collection", "entity")
+        self._container_attributes = dict()
+        super().__init__(*args, **kwargs)
+        self._refresh_container_attributes()
+
+    def _refresh_container_attributes(self):
+        """Refresh values of container properties.
+
+        Store them into private attributes to avoid hitting the database each
+        time we need them.
+        """
+        self._container_attributes = {
+            container_property: getattr(self, f"{container_property}_id", None)
+            for container_property in self._container_properties
+        }
+
+    def set_permission(
+        self, permission: Optional[str], user_or_group: Union[User, Group]
+    ):
+        """Enable use of syntax obj.set_permission instead."""
+        # Resolve circular import.
+        from resolwe.permissions.utils import set_permission
+
+        set_permission(permission, user_or_group, self)
+
+    @property
+    def topmost_container(self) -> Optional[models.Model]:
+        """Get the top-most container.
+
+        :returns: the top-most container or None if it does not exist.
+        """
+        for property in self._container_properties:
+            value = getattr(self, property)
+            if value is not None:
+                return value
+        return None
+
+    def in_container(self, cached: bool = True) -> bool:
+        """Return if object is contained in a container."""
+        if cached:
+            return any(self._container_attributes.values())
+        else:
+            return any(
+                getattr(self, f"{container_property}_id", False)
+                for container_property in self._container_properties
+            )
+
+    def save(self, *args, **kwargs):
+        """Override the save method.
+
+        When 'permission_group' attribute is not set create one. This could
+        lead to a lot unassigned permission groups lying around.
+        """
+        # When object is contaied in the container it inherits the permission
+        # group from the container. The outermost container has precedence.
+        if self.in_container(cached=False):
+            for container_property in self._container_properties:
+                value = getattr(self, f"{container_property}_id", None)
+                # Object was moved to a different container.
+                if value != self._container_attributes[container_property]:
+                    container = getattr(self, container_property)
+                    self.permission_group_id = container.permission_group_id
+                    # Add collection tags to data objects if they are moved to
+                    # a new collection.
+                    if (
+                        self._meta.label in ["flow.Data", "flow.Entity"]
```

Comment:
I don't like this awareness of child classes.

Reply:
Removed.